dvc/remote/gdrive/__init__.py

```diff
@@ -0,0 +1,246 @@
+
+        for part in self.path_info.path.split("/"):
+            if parts and parts[0] == part:
+                parts.pop(0)
+            else:
+                break
+        return parts, [self.root_id]
+
+    def get_path_id_from_cache(self, path_info):
+        files_ids = []
+        parts, parents_ids = self.subtract_root_path(path_info.path.split("/"))
+        if (
+            path_info != self.path_info
+            and parts
+            and (parts[0] in self.cached_dirs)
+        ):
+            parents_ids = self.cached_dirs[parts[0]]
+            files_ids = self.cached_dirs[parts[0]]
+            parts.pop(0)
+
+        return files_ids, parents_ids, parts
+
+    def get_path_id(self, path_info, create=False):
+        files_ids, parents_ids, parts = self.get_path_id_from_cache(path_info)
+
+        if not parts and files_ids:
+            return files_ids[0]
+
+        file1 = self.resolve_remote_file(parents_ids, parts, create)
+        return file1["id"] if file1 else ""
+
+    def exists(self, path_info):
+        return self.get_path_id(path_info) != ""
+
+    def _upload(self, from_file, to_info, name, no_progress_bar):
+        dirname = to_info.parent
+        if dirname:
+            parent_id = self.get_path_id(dirname, True)
+        else:
+            parent_id = to_info.netloc
+
+        upload_request = RequestUploadFile(
+            {
+                "drive": self.drive,
+                "title": to_info.name,
+                "parent_id": parent_id,
+                "mime_type": "",
+            },
+            no_progress_bar,
+            from_file,
+            name,
+        )
+        self.execute_request(upload_request)
+
+    def _download(self, from_info, to_file, name, no_progress_bar):
+        file_id = self.get_path_id(from_info)
+        download_request = RequestDownloadFile(
+            {
+                "drive": self.drive,
+                "file_id": file_id,
+                "to_file": to_file,
+                "progress_name": name,
+                "no_progress_bar": no_progress_bar,
+            }
+        )
+        self.execute_request(download_request)
+
+    def list_cache_paths(self):
+        file_id = self.get_path_id(self.path_info)
+        prefix = self.path_info.path
+        for path in self.list_path(file_id):
+            yield posixpath.join(prefix, path)
+
+    def list_file_path(self, drive_file):
+        if drive_file["mimeType"] == FOLDER_MIME_TYPE:
+            for i in self.list_path(drive_file["id"]):
+                yield posixpath.join(drive_file["title"], i)
+        else:
+            yield drive_file["title"]
+
+    def list_path(self, parent_id):
+        for file1 in self.list_drive_item(
+            "'{}' in parents and trashed=false".format(parent_id)
+        ):
+            for path in self.list_file_path(file1):
+                yield path
+
+    def all(self):
+        if not self.cached_ids:
+            return
+
+        query = " or ".join(
```

Comment:
btw, another question that comes to mind with the way we form all these queries - should we protect from injections at all? or all names are generated by us and we are sure that they do not contain single quotes for example?

Reply:
For the moment, can't see how user's custom data can have impact on queries. Will need to think more about that.
`def all(self):`  lists all files' names by parent folders' ids ( Usage of `id` retrieved from API shouldn't be harmful ). Even if some file names will contain danger data, I guess, DVC will just ignore all file names which are not tracked locally for the moment (and DVC tracks only files names in MD5 format).

Reply:
@MaxRis what about other places? I was more asking about the case if we pass a local file name for some reason into one of these queries - e.g. path to remote. And it contains some special symbols.

Reply:
@shcheklein 

> I was more asking about the case if we pass a local file name for some reason into one of these queries - e.g. path to remote. And it contains some special symbols.

"Custom queries" currently are used only to list gdrive files ( API call doesn't modify content of the remote ). And it seems others PyDrive methods doesn't support "custom query" input param.

Also as of my best knowledge, dvc remote seems operates only with md5 file names already validated/generated before by itself.

Reply:
👍 wanted to make sure